torchmultimodal/modules/layers/parallel_blocks.py

```diff
@@ -0,0 +1,259 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from packaging import version
+from typing import Optional
+import math
+
+from normalizations import RMSNorm
+from position_embedding import RotaryEmbedding
+
+
+def repeat_kv(x: torch.Tensor, n_rep: int) -> torch.Tensor:
+    """expands kv_heads to match q num_heads
+    via
+    torch.repeat_interleave(x, dim=2, repeats=n_rep)"""
+
+    bs, slen, n_kv_heads, head_dim = x.shape
+
+    if n_rep == 1:
+        return x
+    return (
+        x[:, :, :, None, :]
+        .expand(bs, slen, n_kv_heads, n_rep, head_dim)
+        .reshape(bs, slen, n_kv_heads * n_rep, head_dim)
+    )
+
+
+class ParallelAttentionBlock(nn.Module):
+    """
+    Transformer layer multi-head attention and MLP, in a parallelized fashion rather than sequential,
+    with optional attention masking.
+    Inspired by PaLM:  https://arxiv.org/abs/2204.02311
+
+    * We use SwiGLU for the activation function
+    * SwiGLU will approximate same total num params as traditional MLP with GELU
+    * Cross Attention is not enabled here (but available)
+    * MQA and GQA are enabled - modify heads via 'num_heads_group_query_attn'
+    * Bias is enabled by default.  Experiment with removing via use...bias = False, for your application.
+    * Parallel blocks have automated weight initialization via _init_weights.
+    * Please pass in num_layers of your model in num_layers for the weight initialization.
+
+
+    """
+
+    def __init__(
+        self,
+        emb_dimension,
+        num_heads,
+        head_dimension=None,
+        mlp_expansion_ratio: float = 2.6875,  # 8/3 is param matching
+        qk_normalization: bool = True,
+        projection_dropout: float = 0.0,
+        attention_dropout: float = 0.0,
+        use_group_query_attention: bool = True,
+        num_heads_group_query_attention: int = 1,
+        use_in_projection_bias: bool = True,
+        use_out_projection_bias: bool = True,
+        use_weight_init: bool = True,
+        num_layers: int = 1,
+        use_rms_norm: bool = True,
+        use_rotary_embeddings: bool = False,
+        max_expected_seq_len: int = 2048,  # needed only if using rotary
```

Comment:
so if I use rotary embeddings, my sequence length is fixed for pretraining, finetuning, inference? cc @ankitade

Reply:
yes fixed for pretraining. However,  you can do rotary embedding interpolation and expand it though with finetuning (apparently up to 32x).  
https://arxiv.org/pdf/2306.15595.pdf